File: 문제풀이/온라인 저지/프로그래머스/# 02. Level 2/# 068645. 삼각 달팽이/kotlin/Solution.kt
class Solution {
    fun solution(n: Int): IntArray {
        val answer = IntArray((n * (n+1))shr 1)
        var row = 0
        var column = 1
        fun idx(row:Int, column:Int) = ((row * (row - 1)) shr 1) + column - 1
        var number = 1
        var cnt = n
        while (cnt > 0) {
            repeat(cnt) {
                row ++
                answer[idx(row, column)] = number++
            }
            if (--cnt == 0) break
            repeat(cnt) {
                column ++
                answer[idx(row, column)] = number++
            }
            if (--cnt == 0) break
            repeat(cnt) {
                row --
                column --
                answer[idx(row, column)] = number++
            }
            cnt --
        }
        return answer
    }
}

File: 문제풀이/온라인 저지/프로그래머스/# 02. Level 2/# 068645. 삼각 달팽이/python/solution.py
def solution(n):
    answer = [0] * (n * (n + 1) // 2)
    directions = [(1, 0), (0, 1), (-1, -1)]
    r = -1
    c = 0
    max_number = n * (n + 1) // 2
    number = 1

    while number <= max_number:
        for (dr, dc) in directions:
            while number <= max_number:
                nr, nc = r + dr, c + dc
                n_idx = get_idx(nr, nc)
                if (0 <= nr < n) and (0 <= nc <= nr) and answer[n_idx] == 0:
                    answer[n_idx] = number
                    number += 1
                    r, c = nr, nc
                else:
                    break
    return answer

def get_idx(r, c):
    return r * (r + 1) // 2 + c
